src/physics/em/detail/EnergyLossUrbanDistribution.hh

```diff
@@ -0,0 +1,392 @@
+
+    //// DATA ////
+
+    real_type max_energy_;
+    real_type loss_scaling_;
+    Real2     binding_energy_;
+    Real2     xs_exc_;
+    real_type xs_ion_;
+
+    //// CONSTANTS ////
+
+    //! Relative contribution of ionization to energy loss
+    static CELER_CONSTEXPR_FUNCTION real_type rate() { return 0.56; }
+
+    //! Number of collisions above which to use faster sampling from Gaussian
+    static CELER_CONSTEXPR_FUNCTION size_type max_collisions() { return 8; }
+
+    //! Threshold number of excitations used in width correction
+    static CELER_CONSTEXPR_FUNCTION real_type exc_thresh() { return 42; }
+
+    //! Ionization energy [MeV]
+    static CELER_CONSTEXPR_FUNCTION real_type ionization_energy()
+    {
+        return value_as<Energy>(EnergyLossHelper::ionization_energy());
+    }
+
+    //! Energy point below which FWHM scaling doesn't change [MeV]
+    static CELER_CONSTEXPR_FUNCTION real_type fwhm_min_energy()
+    {
+        return 1e-3; // 1 keV
+    }
+
+    //// HELPER FUNCTIONS ////
+
+    template<class Engine>
+    CELER_FUNCTION real_type sample_excitation_loss(Engine& rng);
+
+    template<class Engine>
+    CELER_FUNCTION real_type sample_ionization_loss(Engine& rng);
+
+    template<class Engine>
+    static CELER_FUNCTION real_type sample_fast_urban(real_type mean,
+                                                      real_type stddev,
+                                                      Engine&   rng);
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from distribution parameters.
+ */
+CELER_FUNCTION EnergyLossUrbanDistribution::EnergyLossUrbanDistribution(
+    const FluctuationRef&    shared,
+    const MaterialTrackView& cur_mat,
+    Energy                   unscaled_mean_loss,
+    Energy                   max_energy,
+    Mass                     two_mebsgs,
+    real_type                beta_sq)
+    : max_energy_(max_energy.value())
+{
+    CELER_EXPECT(unscaled_mean_loss > zero_quantity());
+    CELER_EXPECT(two_mebsgs > zero_quantity());
+    CELER_EXPECT(beta_sq > 0);
+
+    // Width correction: the FWHM of the energy loss distribution in thin
+    // layers is in most cases too small; this width correction is used to get
+    // more accurate FWHM values while keeping the mean loss the same by
+    // rescaling the energy levels and number of excitations. The width
+    // correction algorithm is discussed (though not in much detail) in PRM
+    // section 7.3.3
+    loss_scaling_
+        = max_energy_ > this->fwhm_min_energy()
+              ? 1 + real_type(0.5) * this->fwhm_min_energy() / max_energy_
+              : real_type(1.5);
+#if 0
+    // Alternatively...
+    loss_scaling_ = real_type(0.5)
+                        * std::fmin(this->fwhm_min_energy() / max_energy_, 1)
+                    + real_type(1);
+#endif
```

Comment:
Thanks for reminding me. Let me actually write a "cuda snippets" test case here just to see the difference: the first case is what's literally written in geant, to "conditionally" save the division, but the second requires no branching and thus might actually be faster on both GPU/CPU. And I'm not so much concerned about performance as clarity: the second I think shows the bounds more clearly, but I wanted to leave the original in there till the unit tests were complete.

Reply:
So CUDA is "faster": intrinsic "min" and no branching (see https://github.com/celeritas-project/cuda-test-snippets/tree/master/energy-loss-urban). CPU is "slower": external call to `_fmin` or unconditional evaluation of the divisor. I've added fmin(/min/umin)  overloads for CUDA device code, so that CPU will at least inline the "min" function, and optimize for CUDA performance by using its built-ins. This way I don't have to feel bad about substantial CPU performance hits, the GPU version is faster, and most importantly the expression ends up more readable.

Reply:
Excellent!